src/modules/sdbus/test/objpath.c

```diff
@@ -0,0 +1,79 @@
+/************************************************************\
+ * Copyright 2023 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <systemd/sd-bus.h>
+
+#include "src/common/libtap/tap.h"
+#include "ccan/str/str.h"
+#include "ccan/array_size/array_size.h"
+
+#include "objpath.h"
+
+struct testvec {
+    const char *xpath;
+    const char *path;
+};
+
+static struct testvec opvec[] = {
+    { "/object/path/foo.suffix",
+      "/object/path/foo_2esuffix" },
+    { "/org/freedesktop/systemd1/unit/flux-foo.service",
+      "/org/freedesktop/systemd1/unit/flux_2dfoo_2eservice" },
+    { "/foo/flea-bag",
+      "/foo/flea_2dbag" },
+    { "/foo",
+      "/foo" },
+    { "/",
+      "/" },
+};
+
+void test_decode (sd_bus *bus)
+{
+    for (int i = 0; i < ARRAY_SIZE (opvec); i++) {
+        char *p;
+        p = objpath_encode (opvec[i].xpath);
+        diag ("%s", p);
+        ok (p && streq (p, opvec[i].path),
+            "objpath_encode %s works", opvec[i].xpath);
+        free (p);
+        p = objpath_decode (opvec[i].path);
+        ok (p && streq (p, opvec[i].xpath),
+            "objpath_decode %s works", opvec[i].path);
+        free (p);
+    }
+}
+
+int main (int argc, char **argv)
+{
+    sd_bus *bus;
+    int e;
+
+    plan (NO_PLAN);
+
+    if ((e = sd_bus_open_user (&bus)) < 0) {
```

Comment:
i added this into `libsdprocess` unit tests as a sanity check b/c a lot of times they may be run under the developer's user name and not the `flux` user and they may have forgot to set things up.

```
    if (!(dbusenv = getenv ("DBUS_SESSION_BUS_ADDRESS"))
        || !(xdgenv = getenv ("XDG_RUNTIME_DIR"))) {
        diag ("DBUS_SESSION_BUS_ADDRESS or XDG_RUNTIME_DIR not set");
        done_testing ();
    }

    /* Instead of checking if user service running via some equivalent to
     *
     * systemctl list-units user@UID.service
     *
     * just make sure the path to XDG_RUNTIME_DIR exists
     *
     * we don't check DBUS_SESSION_BUS_ADDRESS since it often has
     * funky formatting (e.g. unix:path=/run/user/8556/bus)
     */
    if (access (xdgenv, R_OK | W_OK) < 0) {
        diag ("cannot access XDG_RUNTIME_DIR");
        done_testing ();
    }
```

Reply:
The XDG_RUNTIME_DIR isn't required if DBUS_SESSION_BUS_ADDRESS is set, is it?  See  [source code](https://github.com/systemd/systemd/blob/fde55f3a327c4eef3bbdeb199b05e7c385d1f331/src/libsystemd/sd-bus/sd-bus.c#L1333).  One or the other needs to be set in order for the bus open to succeed, but the tests will still fail if the user session isn't actually running, hence the check.

Was your point that you'd like further hints if the test environment isn't set up right?

Reply:
> Was your point that you'd like further hints if the test environment isn't set up right?

Yup, basically just give extra hints.

You might be right that XDG_RUNTIME_DIR isn't required.  Perhaps it was before?  Or perhaps I just got that from some random guide and have been going with it this whole time.

Reply:
Ah actually, calling systemctl does print hints on my system at least:
```
./t*sdbus*t
build-options:		+systemd+hwloc==2.5.0+zmq==4.3.4
Failed to connect to bus: $DBUS_SESSION_BUS_ADDRESS and $XDG_RUNTIME_DIR not defined (consider using --machine=<user>@.host --user to connect to bus of other user)
1..0 # SKIP user systemd is not running
```